README.md
Magpie
------

Magpie contains a number of scripts for running Big Data software in
HPC environments.  Thus far, Hadoop, Spark, Hbase, Storm, Pig, Mahout,
Phoenix, Kafka, Zeppelin, and Zookeeper are supported.  It currently
supports running over the parallel file system Lustre and running over
any generic network filesytem.  There is scheduler/resource manager
support for Slurm, Moab, Torque, and LSF.

Some of the features presently supported:

- Run jobs interactively or via scripts.
- Run against a number of filesystem options, such as HDFS, HDFS over
  Lustre, HDFS over a generic network filesystem, Lustre directly, or
  a generic network filesystem.
- Take advantage of SSDs/NVRAM for local caching if available
- Run the UDA Infiniband optimization plugin for Hadoop.
- Make decent optimizations for your hardware

Basic Idea
----------

The basic idea behind these scripts are to:

1) Submit a Magpie batch script to allocate nodes on a cluster using
   your HPC scheduler/resource manager.  Slurm, Moab+Slurm,
   Moab+Torque and LSF+mpirun are currently supported.

2) The batch script will create configuration files for all
   appropriate projects (Hadoop, Spark, etc.)  The configuration files
   will be setup so the rank 0 node is the "master".  All compute
   nodes will have configuration files created that point to the node
   designated as the master server.

   The configuration files will be populated with values for your
   filesystem choice and the hardware that exists in your cluster.
   Reasonable attempts are made to determine optimal values for your
   system and hardware (they are almost certainly better than the
   default values).  A number of options exist in the batch scripts to
   adjust these values for individual jobs.

3) Launch daemons on all nodes.  The rank 0 node will run master
   daemons, such as the Hadoop Namenode.  All remaining nodes will run
   appropriate slave daemons, such as the Hadoop Datanodes.

4) Now you have a mini big data cluster to do whatever you want.  You
   can log into the master node and interact with your mini big data
   cluster however you want.  Or you could have Magpie run a script to
   execute your big data calculation instead.

5) When your job completes or your allocation time has run out, Magpie
   will cleanup your job by tearing down daemons.  When appropriate,
   Magpie may also do some additional cleanup work to hopefully make
   re-execution on later runs cleaner and faster.

Supported Packages & Versions
-----------------------------

The following packages and their versions have been tested for minimal
support in this version of Magpie.  Please see full README for
additional details.  Some older versions than those listed here may
still work with Magpie but are not included in Magpie testing.

Versions not listed below should work with Magpie if the
configuration/setup of those versions is compatible with the versions
listed below.  However, certain features or options may not work with
those versions.

* + - Requires patch against binary distro's scripts, no re-compilation needed

Hadoop - 2.2.0+, 2.3.0+, 2.4.0+, 2.4.1+, 2.5.0+, 2.5.1+, 2.5.2+,
         2.6.0+, 2.6.1+, 2.6.2+, 2.6.3+, 2.6.4+, 2.6.5+, 2.7.0+,
         2.7.1+, 2.7.2+, 2.7.3+, 2.8.0+

Spark - 0.9.1-bin-hadoop2+, 0.9.2-bin-hadoop2+,
        1.1.0-bin-hadoop2.3+, 1.1.0-bin-hadoop2.4+,
        1.1.1-bin-hadoop2.3+, 1.1.1-bin-hadoop2.4+,
        1.2.0-bin-hadoop2.3+, 1.2.0-bin-hadoop2.4+,
        1.2.1-bin-hadoop2.3+, 1.2.1-bin-hadoop2.4+,
        1.2.2-bin-hadoop2.3+, 1.2.2-bin-hadoop2.4+,
        1.3.0-bin-hadoop2.3+, 1.3.0-bin-hadoop2.4+,
        1.3.1-bin-hadoop2.3+, 1.3.1-bin-hadoop2.4+,
        1.3.1-bin-hadoop2.6+, 1.4.0-bin-hadoop2.3+,
        1.4.0-bin-hadoop2.4+, 1.4.0-bin-hadoop2.6+,
        1.4.1-bin-hadoop2.3+, 1.4.1-bin-hadoop2.4+,
        1.4.1-bin-hadoop2.6+, 1.5.0-bin-hadoop2.6+,
        1.5.1-bin-hadoop2.6+, 1.5.2-bin-hadoop2.6+,
        1.6.0-bin-hadoop2.6+, 1.6.1-bin-hadoop2.6+,
        1.6.2-bin-hadoop2.6+, 1.6.3-bin-hadoop2.6+,
        2.0.0-bin-hadoop2.6+, 2.0.0-bin-hadoop2.7+,
        2.0.1-bin-hadoop2.6+, 2.0.1-bin-hadoop2.7+,
        2.0.2-bin-hadoop2.6+, 2.0.2-bin-hadoop2.7+,
        2.1.0-bin-hadoop2.6+, 2.1.0-bin-hadoop2.7+,
        2.1.1-bin-hadoop2.6+, 2.1.1-bin-hadoop2.7+

Hbase - 0.98.0-hadoop2+, 0.98.1-hadoop2+, 0.98.2-hadoop2+,
        0.98.3-hadoop2+, 0.98.4-hadoop2+, 0.98.5-hadoop2+,
        0.98.6-hadoop2+, 0.98.6.1-hadoop2+, 0.98.7-hadoop2+,
        0.98.8-hadoop2+, 0.98.9-hadoop2+, 0.98.10-hadoop2+,
        0.98.10.1-hadoop2+, 0.98.11-hadoop2+, 0.98.12-hadoop2+,
        0.98.12.1-hadoop2+, 0.98.13-hadoop2+, 0.98.14-hadoop2+,
        0.98.15-hadoop2+, 0.98.16-hadoop2+, 0.98.16.1-hadoop2+,
        0.98.17-hadoop2+, 0.98.18-hadoop2+, 0.98.19-hadoop2+,
        0.98.20-hadoop2+, 0.98.21-hadoop2+, 0.98.22-hadoop2+,
        0.98.23-hadoop2+, 0.98.24-hadoop2+, 0.99.0+, 0.99.1+, 0.99.2+,
        1.0.0+, 1.0.1+, 1.0.1.1+, 1.0.2+, 1.0.3+, 1.1.0+, 1.1.0.1+,
        1.1.1+, 1.1.2+, 1.1.3+, 1.1.4+, 1.1.5+, 1.1.6+, 1.1.7+,
        1.1.8+, 1.1.9+, 1.2.0+, 1.2.1+, 1.2.2+, 1.2.3+, 1.2.4+,
        1.2.5+, 1.3.0+, 1.3.1+

Pig - 0.13.0, 0.14.0, 0.15.0, 0.16.0

Zookeeper - 3.4.0, 3.4.1, 3.4.2, 3.4.3, 3.4.4, 3.4.5, 3.4.6, 3.4.7,
            3.4.8, 3.4.9

Storm - 0.9.3, 0.9.4, 0.9.5, 0.9.6, 0.9.7, 0.10.0, 0.10.1, 0.10.2,
        1.0.0, 1.0.1, 1.0.2, 1.0.3, 1.1.0

Mahout - 0.11.0+, 0.11.1+, 0.11.2+, 0.12.0+, 0.12.1+, 0.12.2, 0.13.0

Phoenix - 4.5.0-Hbase-1.0+, 4.5.0-Hbase-1.1+, 4.5.1-Hbase-1.0+,
          4.5.1-Hbase-1.1+, 4.5.2-HBase-1.0+, 4.5.2-HBase-1.1+,
          4.6.0-Hbase-1.0+, 4.6.0-Hbase-1.1, 4.7.0-Hbase-1.0+,
          4.7.0-Hbase-1.1, 4.8.0-Hbase-1.0+, 4.8.0-Hbase-1.1,
          4.8.0-Hbase-1.2, 4.8.1-Hbase-1.0+, 4.8.1-Hbase-1.1,
          4.8.1-Hbase-1.2, 4.8.2-Hbase-1.0+, 4.8.2-Hbase-1.1,
          4.8.2-Hbase-1.2, 4.9.0-Hbase-1.1, 4.9.0-Hbase-1.2,
          4.10.0-Hbase-1.1, 4.10.0-Hbase-1.2

Kafka - 2.11-0.9.0.0

Zeppelin - 0.6.0, 0.6.1, 0.6.2, 0.7.0, 0.7.1

Older Supported Packages & Features
-----------------------------------

Some packages and features were dropped in Magpie 2.0 due to lack of
interest, the software becoming old/deprecated, and/or their initial
experimental addition into Magpie.  If you are interested in them,
please look at Magpie 1.X versions for supported versions and
documentation.  If you are very interested in support in Magpie 2.0
beyond an experimental nature, please submit a support request and we
can reconsider adding it back in.

   - Hadoop 1.X support
   - Tachyon
   - UDA/uda-plugin for Hadoop
   - HDFS Federation in Hadoop
   - IntelLustre option for a Hadoop Filesystem
   - MagpieNetworkFS option for a Hadoop Filesystem

Documentation
-------------

All documentation is in the 'doc' subdirectory.  Please see the
doc/README file as a starting point.  It provides general instructions
as well as pointers to documentation for each project, setup
requirements, ability to do local configurations, tips & tricks, and
more information.

Release
-------

Magpie is release under a GPL license. For more information, see the [COPYING](/COPYING) file.

`LLNL-CODE-644248`


NEWS
Magpie 2.0
----------

Magpie 2.0 includes new software support, but primarily involves a
large amount of re-architecture and removal of legacy experimental
support.  Job submission scripts from Magpie 1.X versions will be
incompatible with Magpie 2.0.  So please read details below for
highlights.

New Support

o Support Hadoop version 3

  - Support new HADOOP_WORKER_COUNT & HADOOP_WORKER_CORE_COUNT
    environment variables.

  - Update default Hadoop version to 3.0.0-alpha2-SNAPSHOT (XXX).

Re-architecture and changes

o Rearchitect MAGPIE_JOB_TYPE to be overarching job configuration
  variable.

  - "setuponly" and "interactive" jobs are now configured by
    MAGPIE_JOB_TYPE and not through any of the former "MODE" variables
    (e.g. HADOOP_MODE, SPARK_MODE, etc.)

  - Generic "script" jobs are now configured by MAGPIE_JOB_TYPE and
    not through any of the former "MODE" variables.
    (e.g. HADOOP_MODE, SPARK_MODE, etc.)

    However, "script" is still available for those projects that run a
    script through a specific service, e.g. PIG and PHOENIX.

o Rearchitect HADOOP_SETUP_TYPE to be version agnostic.  It now takes
  "MR", "YARN", or "HDFS" as inputs.

  - A new submission script variant to support Spark w/ Yarn and w/o
    HDFS is now available.

o "MODE" variables (e.g. HADOOP_MODE, SPARK_MODE, etc.) are now named
  "JOB" (e.g. HADOOP_JOB, SPARK_JOB, etc.).  This change was done to
  indicate that they truly only run jobs and are not meant for
  configuring the project in any way.

  - As a result the following configurations have been removed.

    "launch" from HADOOP_JOB
    "hdfsonly" from HADOOP_JOB
    "server" from KAFKA_JOB
    "server" from PHOENIX_JOB
    "launch" from ZOOKEEPER_JOB
    "server" from ZEPPELIN_JOB

  - The Phoenix mode "server" is now available through the
    configuration variable PHOENIX_START_QUERYSERVER.

o SPARK_USE_YARN has been removed and replaced with SPARK_SETUP_TYPE
  with settings options of "STANDALONE" or "YARN".

o Remove experimental features and experimental/non-heavily used
  projects.  If you wish for these features to be brought back beyond
  an 'experimental' nature, please submit an issue request.  Features
  & projects listed below.

  - Hadoop 1.X versions
  - Tachyon
  - UDA
  - 'MagpieNetworkFS' Filesystem
  - 'IntelLustre' Filesystem
  - HDFS Federation

o Older "quirky" version numbers in Hadoop (e.g. 0.21.0 in Hadoop) are
  no longer supported.

Misc Features & Changes

- MAGPIE_PRE_JOB_RUN and MAGPIE_POST_JOB_RUN can have multiple scripts
  listed by comma.

- Default HDFS namenode port is now 8020 instead of 54310

- Update default Java to 1.8 in submission scripts.

Misc Documentation Changes

- No longer list support for versions that require re-compilation.

- No longer list support for versions that are not included in Magpie
  testsuite.

Backwards Compatibility Notes

o The following are possible changes users may need to make in older
  job submission scripts.

  - "MODE" variables (e.g. HADOOP_MODE, SPARK_MODE, etc.) are now
    named "JOB" (e.g. HADOOP_JOB, SPARK_JOB, etc.).

  - "JOB" variables no longer support "interactive" or "setuponly"
    options.  These are now controlled via MAGPIE_JOB_TYPE.

  - "JOB" variables no longer support "server", "launch", or
    "hdfsonly" options.  A service "launching" is always assumed
    unless MAGPIE_JOB_TYPE is set to "setuponly".

  - Many "JOB" variables no longer support "script" options, these are
    done via MAGPIE_JOB_TYPE.

  - "PHOENIX_MODE"'s "server" option is now configured through
    PHOENIX_START_QUERYSERVER.

  - HADOOP_SETUP_TYPE no longer takes input of "MR1", "MR2", "HDFS1",
    or "HDFS2".  It now only takes "MR", "YARN", or "HDFS".

  - SPARK_USE_YARN is now configured through SPARK_SETUP_TYPE.

  - HADOOP_FILESYSTEM_MODE no longer takes input of "magpienetworkfs"
    or "intellustre".

magpie 1.82
-----------
Add support for Hbase 1.3.1
- Update default Hbase to 1.3.1
Add support for Spark 2.1.1-hadoop2.6, Spark 2.1.1-hadoop2.7.
- Update default Spark to 2.1.1-hadoop2.7.
Add support for Mahout 0.13.0.
- Update default Mahout to 0.13.0

magpie 1.81
-----------
Update lfs usage to use --stripe-size over deprecated --size option.
  lfs will no longer work with Lustre versions prior to 2.2.51.
Add support for Hadoop 2.8.0
- Update default Hadoop to 2.8.0.
Add support for Zeppelin 0.7.0, 0.7.1.
- Update default Zeppelin to 0.7.1.
Add support for Storm 0.9.7, 1.0.3, 1.1.0.
- Update default Storm to 1.1.0.
Add support for Hbase 1.1.9, 1.2.5.
Add support for Phoenix 4.10.0-Hbase-1.1, 4.10.0-Hbase-1.2.
- Update default Phoenix to 4.10.0-Hbase-1.2.
Increase default
  yarn.nodemanager.disk-health-checker.max-disk-utilization-per-disk-percentage
  in Hadoop to 98%.
Fix bug in calculation of MAGPIE_CLUSTER_NODERANK for submission type
  'lsfmpirun'.
When calculating user file descriptor limits and process limits to
  use, do not decrease the number if the system provides a much higher
  default.

magpie 1.80
-----------
Fix bug in which all script ARGS variables were not passed to scripts
  that were executed.
Add support for Spark 2.1.0-hadoop2.6, Spark 2.1.0-hadoop2.7.
- Update default Spark to 2.1.0-hadoop2.7. 
Add support for Hbase 1.3.0
- Update default Hbase to 1.3.0
Add support for Hbase 0.98.24
Update default Spark used w/ Zeppelin to 1.6.3-bin-hadoop2.6.

magpie 1.79
-----------
Fix corner case in job name into paths for HADOOP_PER_JOB_HDFS_PATH or
  ZOOKEEPER_PER_JOB_DATA_DIR from 1.78.
Add support for Spark 2.0.2-hadoop2.6, Spark 2.0.2-hadoop2.7.
- Update default Spark to 2.0.2-hadoop2.7. 
Add support for Spark 1.6.3.
Add support for Hbase 1.2.4.
- Update default Hbase to 1.2.4. 
Add support for Hbase 1.1.7, 1.1.8.
Add support for Hbase 0.98.23
Add support for Phoenix 4.8.2-Hbase-1.0, 4.8.2-Hbase-1.1,
  4.8.2-Hbase-1.2, 4.9.0-Hbase-1.1, 4.9.0-Hbase-1.2.
- Update default Phoenix to 4.9.0-Hbase-1.2.

magpie 1.78
-----------

testsuite/test-generate-common.sh
#!/bin/bash

source test-common.sh
source test-config.sh
source ../magpie/lib/magpie-lib-helper

GetMinutesJob () {
    local addminutes=$1
    __minutesjob=`expr $STARTUP_TIME + $SHUTDOWN_TIME + $addminutes`
    timeoutputforjob=$__minutesjob
}

GetSecondsJob () {
    local addminutes=$1
    GetMinutesJob $addminutes
    secondsjob=`expr $__minutesjob \* 60`
    timeoutputforjob=$secondsjob
}

GetHoursMinutesJob () {
    local addminutes=$1
    GetMinutesJob $1
    local hours=`expr $__minutesjob / 60`
    local minutesleft=`expr $__minutesjob % 60`
    timeoutputforjob=`printf "%d:%02d" ${hours} ${minutesleft}`
}

SetupMagpieTestScript () {
    local testscript=$1
    shift
    local files=$@

    sed -i \
        -e 's/export MAGPIE_JOB_TYPE="\(.*\)"/export MAGPIE_JOB_TYPE="script"/' \
        -e 's/# export MAGPIE_SCRIPT_PATH="\(.*\)"/export MAGPIE_SCRIPT_PATH="'"${magpiescriptshomesubst}"'\/testsuite\/testscripts\/'"${testscript}"'"/' \
        ${files}
}

JavaCommonSubstitution() {
    local javaversion=$1

    shift
    local files=$@

    if [ "${javaversion}" == "${java16}" ]
    then
        sed -i -e 's/export JAVA_HOME="\(.*\)"/export JAVA_HOME="'"${java16pathsubst}"'"/' ${files}
    elif [ "${javaversion}" == "${java17}" ]
    then
        sed -i -e 's/export JAVA_HOME="\(.*\)"/export JAVA_HOME="'"${java17pathsubst}"'"/' ${files}
    elif [ "${javaversion}" == "${java18}" ]
    then
        sed -i -e 's/export JAVA_HOME="\(.*\)"/export JAVA_HOME="'"${java18pathsubst}"'"/' ${files}
    else
        echo "Invalid Java Version Specified - ${javaversion}"
        exit 1
    fi
}

CheckForDependency() {
    local project=$1
    local projectcheck=$2
    local checkversion=$3

    local checkversionunderscore=`echo ${checkversion} | sed -e "s/\./_/g" -e "s/-/_/g"`
    local projectchecklowercase=`echo ${projectcheck} | tr '[:upper:]' '[:lower:]'`
    local variabletocheck="${projectchecklowercase}_${checkversionunderscore}"

    if [ "${!variabletocheck}" != "y" ]
    then
        echo "Cannot generate ${project} tests that depend on ${projectcheck} ${checkversion}, it's not enabled"
        continue
    fi
}

RemoveTestsCheck() {
    local project=$1
    local version=$2

    local versionunderscore=`echo ${version} | sed -e "s/\./_/g" | sed -e "s/-/_/g"`
    local projectlowercase=`echo ${project} | tr '[:upper:]' '[:lower:]'`
    local variabletocheck="${projectlowercase}_${versionunderscore}"

    if [ "${!variabletocheck}" == "n" ]
    then
        rm -f magpie.${submissiontype}*${project}-${version}*
    fi
}

testsuite/test-common.sh
#!/bin/bash

java16="1.6"
java17="1.7"
java18="1.8"

hadoop_decomissionhdfs_minimum="2.3.0"

hadoop2Xjava16versions="2.2.0 2.3.0 2.4.0 2.4.1 2.5.0 2.5.1 2.5.2"
hadoop2Xjava16versions_javaversion=${java16}
hadoop2Xjava17versions="2.6.0 2.6.1 2.6.2 2.6.3 2.6.4 2.6.5 2.7.0 2.7.1 2.7.2 2.7.3 2.8.0"
hadoop2Xjava17versions_javaversion=${java17}
hadoop3Xjava17versions="3.0.0-alpha2-SNAPSHOT"
hadoop3Xjava17versions_javaversion=${java18}

hadoop_test_groups="hadoop2Xjava16versions hadoop2Xjava17versions hadoop3Xjava17versions"
hadoop_all_versions="${hadoop2Xjava16versions} ${hadoop2Xjava17versions} ${hadoop3Xjava17versions}"

mahouthadoop27java17versions="0.11.0 0.11.1 0.11.2 0.12.0 0.12.1 0.12.2 0.13.0"
mahouthadoop27java17versions_hadoopversion="2.7.0"
mahouthadoop27java17versions_javaversion=${java17}

mahout_test_groups="mahouthadoop27java17versions"
mahout_all_versions="${mahouthadoop27java17versions}"

pighadoop26java16versions="0.13.0 0.14.0"
pighadoop26java16versions_hadoopversion="2.6.0"
pighadoop26java16versions_javaversion=${java16}
pighadoop27java17versions="0.15.0 0.16.0"
pighadoop27java17versions_hadoopversion="2.7.0"
pighadoop27java17versions_javaversion=${java17}

pig_test_groups="pighadoop26java16versions pighadoop27java17versions"
pig_all_versions="${pighadoop26java16versions} ${pighadoop27java17versions}"

hbasehadoop22zookeeper34java16versions="0.98.0-hadoop2 0.98.1-hadoop2 0.98.2-hadoop2 0.98.3-hadoop2 0.98.4-hadoop2 0.98.5-hadoop2 0.98.6-hadoop2 0.98.6.1-hadoop2 0.98.7-hadoop2 0.98.8-hadoop2 0.98.9-hadoop2 0.98.10-hadoop2 0.98.10.1-hadoop2 0.98.11-hadoop2 0.98.12-hadoop2 0.98.12.1-hadoop2 0.98.13-hadoop2 0.98.14-hadoop2 0.98.15-hadoop2 0.98.16-hadoop2 0.98.16.1-hadoop2 0.98.17-hadoop2 0.98.18-hadoop2 0.98.19-hadoop2 0.98.20-hadoop2 0.98.21-hadoop2 0.98.22-hadoop2 0.98.23-hadoop2 0.98.24-hadoop2"
hbasehadoop22zookeeper34java16versions_hadoopversion="2.2.0"
hbasehadoop22zookeeper34java16versions_zookeeperversion="3.4.6"
hbasehadoop22zookeeper34java16versions_javaversion=${java16}
hbasehadoop27zookeeper34java17versions="0.99.0 0.99.1 0.99.2 1.0.0 1.0.1 1.0.1.1 1.0.2 1.0.3 1.1.0 1.1.0.1 1.1.1 1.1.2 1.1.3 1.1.4 1.1.5 1.1.6 1.1.7 1.1.8 1.1.9 1.2.0 1.2.1 1.2.2 1.2.3 1.2.4 1.2.5 1.3.0 1.3.1"
hbasehadoop27zookeeper34java17versions_hadoopversion="2.7.0"
hbasehadoop27zookeeper34java17versions_zookeeperversion="3.4.9"
hbasehadoop27zookeeper34java17versions_javaversion=${java17}

hbase_test_groups="hbasehadoop22zookeeper34java16versions hbasehadoop27zookeeper34java17versions"
hbase_all_versions="${hbasehadoop22zookeeper34java16versions} ${hbasehadoop27zookeeper34java17versions}"

phoenixhbase10hadoop27zookeeper34java17versions="4.5.0-HBase-1.0 4.5.1-HBase-1.0 4.5.2-HBase-1.0 4.6.0-HBase-1.0 4.7.0-HBase-1.0 4.8.0-HBase-1.0 4.8.1-HBase-1.0 4.8.2-HBase-1.0"
phoenixhbase10hadoop27zookeeper34java17versions_hbaseversion="1.0.0"
phoenixhbase10hadoop27zookeeper34java17versions_hadoopversion="2.7.0"
phoenixhbase10hadoop27zookeeper34java17versions_zookeeperversion="3.4.9"
phoenixhbase10hadoop27zookeeper34java17versions_javaversion=${java17}

phoenixhbase11hadoop27zookeeper34java17versions="4.5.0-HBase-1.1 4.5.1-HBase-1.1 4.5.2-HBase-1.1 4.6.0-HBase-1.1 4.7.0-HBase-1.1 4.8.0-HBase-1.1 4.8.1-HBase-1.1 4.8.2-HBase-1.1 4.9.0-HBase-1.1 4.10.0-HBase-1.1"
phoenixhbase11hadoop27zookeeper34java17versions_hbaseversion="1.1.0"
phoenixhbase11hadoop27zookeeper34java17versions_hadoopversion="2.7.0"
phoenixhbase11hadoop27zookeeper34java17versions_zookeeperversion="3.4.9"
phoenixhbase11hadoop27zookeeper34java17versions_javaversion=${java17}

phoenixhbase12hadoop27zookeeper34java17versions="4.8.0-HBase-1.2 4.8.1-HBase-1.2 4.8.2-HBase-1.2 4.9.0-HBase-1.2 4.10.0-HBase-1.2"
phoenixhbase12hadoop27zookeeper34java17versions_hbaseversion="1.2.0"
phoenixhbase12hadoop27zookeeper34java17versions_hadoopversion="2.7.0"
phoenixhbase12hadoop27zookeeper34java17versions_zookeeperversion="3.4.9"
phoenixhbase12hadoop27zookeeper34java17versions_javaversion=${java17}

phoenix_test_groups="phoenixhbase10hadoop27zookeeper34java17versions phoenixhbase11hadoop27zookeeper34java17versions phoenixhbase12hadoop27zookeeper34java17versions"
phoenix_all_versions="${phoenixhbase10hadoop27zookeeper34java17versions} ${phoenixhbase11hadoop27zookeeper34java17versions} ${phoenixhbase12hadoop27zookeeper34java17versions}"

spark0Xjava16hadoop2versions="0.9.1-bin-hadoop2 0.9.2-bin-hadoop2"
spark0Xjava16hadoop2versions_hadoopversion="2.2.0"
spark0Xjava16hadoop2versions_javaversion=${java16}
spark1Xjava16hadoop23versions="1.1.0-bin-hadoop2.3 1.1.1-bin-hadoop2.3 1.2.0-bin-hadoop2.3 1.2.1-bin-hadoop2.3 1.2.2-bin-hadoop2.3 1.3.0-bin-hadoop2.3 1.3.1-bin-hadoop2.3 1.4.0-bin-hadoop2.3 1.4.1-bin-hadoop2.3"
spark1Xjava16hadoop23versions_hadoopversion="2.3.0"
spark1Xjava16hadoop23versions_javaversion=${java16}
spark1Xjava16hadoop24versions="1.1.0-bin-hadoop2.4 1.1.1-bin-hadoop2.4 1.2.0-bin-hadoop2.4 1.2.1-bin-hadoop2.4 1.2.2-bin-hadoop2.4 1.3.0-bin-hadoop2.4 1.3.1-bin-hadoop2.4 1.4.0-bin-hadoop2.4 1.4.1-bin-hadoop2.4"
spark1Xjava16hadoop24versions_hadoopversion="2.4.0"
spark1Xjava16hadoop24versions_javaversion=${java16}
spark1Xjava17hadoop26versions="1.3.1-bin-hadoop2.6 1.4.0-bin-hadoop2.6 1.4.1-bin-hadoop2.6 1.5.0-bin-hadoop2.6 1.5.1-bin-hadoop2.6 1.5.2-bin-hadoop2.6 1.6.0-bin-hadoop2.6 1.6.1-bin-hadoop2.6 1.6.2-bin-hadoop2.6"
spark1Xjava17hadoop26versions_hadoopversion="2.6.0"
spark1Xjava17hadoop26versions_javaversion=${java17}
spark2Xjava17hadoop26versions="2.0.0-bin-hadoop2.6 2.0.1-bin-hadoop2.6 2.0.2-bin-hadoop2.6 2.1.0-bin-hadoop2.6 2.1.1-bin-hadoop2.6"
spark2Xjava17hadoop26versions_hadoopversion="2.6.0"
spark2Xjava17hadoop26versions_javaversion=${java17}
spark2Xjava17hadoop27versions="2.0.0-bin-hadoop2.7 2.0.1-bin-hadoop2.7 2.0.2-bin-hadoop2.7 2.1.0-bin-hadoop2.7 2.1.1-bin-hadoop2.7"
spark2Xjava17hadoop27versions_hadoopversion="2.7.0"
spark2Xjava17hadoop27versions_javaversion=${java17}

spark_test_groups="spark0Xjava16hadoop2versions spark1Xjava16hadoop23versions spark1Xjava16hadoop24versions spark1Xjava17hadoop26versions spark2Xjava17hadoop26versions spark2Xjava17hadoop27versions"
spark_all_versions="${spark0Xjava16hadoop2versions} ${spark1Xjava16hadoop23versions} ${spark1Xjava16hadoop24versions} ${spark1Xjava17hadoop26versions} ${spark2Xjava17hadoop26versions} ${spark2Xjava17hadoop27versions}"

stormzookeeper34java16versions="0.9.3 0.9.4"
stormzookeeper34java16versions_zookeeperversion="3.4.6"
stormzookeeper34java16versions_javaversion=${java16}
stormzookeeper34java17versions="0.9.5 0.9.6 0.9.7 0.10.0 0.10.1 0.10.2 1.0.0 1.0.1 1.0.2 1.0.3 1.1.0"
stormzookeeper34java17versions_zookeeperversion="3.4.9"
stormzookeeper34java17versions_javaversion=${java17}

storm_test_groups="stormzookeeper34java16versions stormzookeeper34java17versions"
storm_all_versions="${stormzookeeper34java16versions} ${stormzookeeper34java17versions}"

kafkazookeeper34java17versions="2.11-0.9.0.0"
kafkazookeeper34java17versions_zookeeperversion="3.4.9"
kafkazookeeper34java17versions_javaversion=${java17}

kafka_test_groups="kafkazookeeper34java17versions"
kafka_all_versions="${kafkazookeeper34java17versions}"

zookeeperjava17versions="3.4.0 3.4.1 3.4.2 3.4.3 3.4.4 3.4.5 3.4.6 3.4.7 3.4.8 3.4.9"
zookeeperjava17versions_javaversion=${java17}

zookeeper_test_groups="zookeeperjava17versions"
zookeeper_all_versions="${zookeeperjava17versions}"

zeppelinspark16java17versions="0.6.0 0.6.1 0.6.2 0.7.0 0.7.1"
zeppelinspark16java17versions_sparkversion="1.6.0-bin-hadoop2.6"
zeppelinspark16java17versions_javaversion="${java17}"

zeppelin_test_groups="zeppelinspark16java17versions"
zeppelin_all_versions="${zeppelinspark16java17versions}"
